File: p140/p140.cpp
// Memory Search

class Solution {
private:
    unordered_map<string, vector<string>> m;
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());

        vector<string> tmp;
        m[""] = tmp;

        dfs(s, dict);

        return m[s];
    }

    void dfs(string s, unordered_set<string> &dict)
    {
        if(m.count(s))
            return;

        vector<string> res;
        if (dict.count(s))
            res.push_back(s);

        for (int i = 1; i < s.length(); ++i)
        {
            string pre = s.substr(0, i);
            if (!dict.count(pre))
                continue;

            string rem = s.substr(i, s.length() - i);
            dfs(rem, dict);
            for (int j = 0; j < m[rem].size(); ++j)
            {
                string tmp = pre + " " + m[rem][j];
                res.push_back(tmp);
            }
        }

        m[s] = res;
    }
};

File: p140/p140.go
// Memory Search. 4ms. Same algorithm with the C++ version, but much faster.
func wordBreak(s string, wordDict []string) []string {
    m := make(map[string][]string)

    dict := make(map[string]bool)
    for _, v := range wordDict {
    	dict[v] = true
    }

    m[""] = []string{""}

    dfs(s, &m, &dict)

    return m[s]
}

func dfs(s string, m *map[string][]string, dict *map[string]bool) {
	_, ok := (*m)[s]
	if ok {
		return
	}

	(*m)[s] = []string{}
	if (*dict)[s] == true {
		(*m)[s] = append((*m)[s], s)
	}

	n := len(s)
	for i := 1; i < n; i++ {
		pre := s[:i]

		if (*dict)[pre] == false {
			continue
		}

		rem := s[i:]
		dfs(rem, m, dict)

		for _, suffix := range (*m)[rem] {
			(*m)[s] = append((*m)[s], pre + " " + suffix)
		}
	}
}